ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -828,26 +829,30 @@ func (r *RayClusterReconciler) reconcilePods(ctx context.Context, instance *rayv
 // (1) shouldDelete: Whether the Pod should be deleted.
 // (2) reason: The reason why the Pod should or should not be deleted.
 func shouldDeletePod(pod corev1.Pod, nodeType rayv1.RayNodeType) (bool, string) {
-	// If a Pod's restart policy is set to `Always`, KubeRay will not delete
-	// the Pod and rely on the Pod's restart policy to restart the Pod.
+	// If a Pod's restart policy is set to `Always` and the Pod is not evicted,
+	// KubeRay will not delete the Pod and rely on the Pod's restart policy to restart the Pod.
 	isRestartPolicyAlways := pod.Spec.RestartPolicy == corev1.RestartPolicyAlways
+	isPodEvicted := pod.Status.Reason == utils.PodEvictedReason
```

Comment:
Are we sure pod eviction is the only case where pod phase is Failed? Should we future proof this logic for future cases and always delete pod when phase is Failed? 
```go
func shouldDeletePod(pod corev1.Pod, nodeType rayv1.RayNodeType) (bool, string) {
	// If the Pod's status is `Failed` or `Succeeded`, the Pod will not restart and we can safely delete it.
	if pod.Status.Phase == corev1.PodFailed || pod.Status.Phase == corev1.PodSucceeded {
	    // A Pod with `RestartPolicy: Always` will guarantee that the status transition
	    // from `Running` to `Failed / Succeeded` and back to `Running`.
	    // However, if the Non-Controller Managed Pod(e.g., head pod) is evicted
	    // because of external factors (e.g., node failure, insufficient resources, etc.),
	    // then the Pod will not be restarted and remain in the `Failed` status indefinitely.
	    // So we need to delete them
	    reason := fmt.Sprintf(
		"The %s Pod %s status is %s which is a terminal state "+
			"KubeRay will delete the Pod and create new Pods in the next reconciliation if necessary.",
			nodeType, pod.Name, pod.Status.Phase)
	   return true, reason
	}

       ...
       ...
```

Reply:
Eviction is not the only reason for phase `Failed`. Currently managed pods will briefly fail and quickly transition to running (restart policy Always) and we no need to delete it, but if Eviction they will not and keep `Failed`.
So handle this scene separately

Reply:
> Currently managed pods will briefly fail and quickly transition to running (restart policy Always)

The existing comments here imply that this is not the case: https://github.com/ray-project/kuberay/blob/8e5f88ca35cd0c6a4de4162ef4dda83c81515275/ray-operator/controllers/ray/raycluster_controller.go#L838-L841 

Have you tested when restart policy is Always and the reason is not Evicted, that the phase temporarily changes to `Failed`?

Reply:
Thank you for your question. I have re examined and found that my previous understanding may have been incorrect. This is the logic I found about pod status updates from the k8s source code, and I found that if the restart strategy remains always, its status remains running
https://github.com/kubernetes/kubernetes/blob/46aa8959a0659e22c924bb52b38385d441715b2b/pkg/kubelet/kubelet_pods.go#L1556

For your scenario, there should not be such a situation, so it cannot be reproduced.
However, from Failed to Running, there is indeed a possibility of human intervention in pod configuration, node resource recovery, etc. Therefore, is it necessary to handle this concurrent scenario: when a node fails, it also needs to query the latest node status and perform a deletion operation